Python/import.c

```diff
@@ -386,12 +386,115 @@ import_get_module(PyThreadState *tstate, PyObject *name)
+            interp->importlib, &PyId__lock_unlock_module, name);
+        if (value == NULL) {
+            Py_DECREF(spec);
+            return -1;
+        }
+        else {
+            return 0;
+        }
+        Py_DECREF(value);
+    }
+    Py_XDECREF(spec);
+    return 0;
+}
+
+
+/* Remove importlib frames from the traceback,
+ * except in Verbose mode. */
+static void
+remove_importlib_frames(PyThreadState *tstate)
+{
+    const char *importlib_filename = "<frozen importlib._bootstrap>";
+    const char *external_filename = "<frozen importlib._bootstrap_external>";
+    const char *remove_frames = "_call_with_frames_removed";
+    int always_trim = 0;
+    int in_importlib = 0;
+    PyObject *exception, *value, *base_tb, *tb;
+    PyObject **prev_link, **outer_link = NULL;
+
+    /* Synopsis: if it's an ImportError, we trim all importlib chunks
+       from the traceback. We always trim chunks
+       which end with a call to "_call_with_frames_removed". */
+
+    _PyErr_Fetch(tstate, &exception, &value, &base_tb);
+    if (!exception || tstate->interp->config.verbose) {
+        goto done;
+    }
+
+    if (PyType_IsSubtype((PyTypeObject *) exception,
+                         (PyTypeObject *) PyExc_ImportError))
+        always_trim = 1;
+
+    prev_link = &base_tb;
+    tb = base_tb;
+    while (tb != NULL) {
+        PyTracebackObject *traceback = (PyTracebackObject *)tb;
+        PyObject *next = (PyObject *) traceback->tb_next;
+        PyFrameObject *frame = traceback->tb_frame;
+        PyCodeObject *code = frame->f_code;
+        int now_in_importlib;
+
+        assert(PyTraceBack_Check(tb));
+        now_in_importlib = _PyUnicode_EqualToASCIIString(code->co_filename, importlib_filename) ||
+                           _PyUnicode_EqualToASCIIString(code->co_filename, external_filename);
+        if (now_in_importlib && !in_importlib) {
+            /* This is the link to this chunk of importlib tracebacks */
+            outer_link = prev_link;
+        }
+        in_importlib = now_in_importlib;
+
+        if (in_importlib &&
+            (always_trim ||
+             _PyUnicode_EqualToASCIIString(code->co_name, remove_frames))) {
+            Py_XINCREF(next);
+            Py_XSETREF(*outer_link, next);
+            prev_link = outer_link;
+        }
+        else {
+            prev_link = (PyObject **) &traceback->tb_next;
+        }
+        tb = next;
+    }
+done:
+    _PyErr_Restore(tstate, exception, value, base_tb);
+}
+
 
 PyObject *
 PyImport_GetModule(PyObject *name)
 {
-    PyThreadState *tstate = _PyThreadState_GET();
-    return import_get_module(tstate, name);
+   PyThreadState *tstate = _PyThreadState_GET();
+   PyObject *mod = NULL;
+
+   mod = import_get_module(tstate, name);
+   if (mod != NULL && mod != Py_None) {
+       int value = import_ensure_initialized(tstate, mod, name);
+       if (value == -1) {
+            remove_importlib_frames(tstate);
+            return NULL;
```

Comment:
I'm looking back at what the `goto error` in the original context does, as a guide to what this error case should likely do.

One is `remove_importlib_frames` -- great. A couple of other steps there don't appear to apply.

But another one is `Py_XDECREF(mod);`. And in fact it does look like we should do that here. (If there hadn't been an error, we'd have returned `mod` to the caller -- which must mean the caller would own the reference, so it must be that we own the reference here. Which means if we return `NULL` and so *don't* pass the pointer on to some other code that will own the reference, it's up to us here to decref it away.)

I think that's then all of it.

Reply:
> 
> 
> I'm looking back at what the `goto error` in the original context does, as a guide to what this error case should likely do.
> 
> One is `remove_importlib_frames` -- great. A couple of other steps there don't appear to apply.
> 
> But another one is `Py_XDECREF(mod);`. And in fact it does look like we should do that here. (If there hadn't been an error, we'd have returned `mod` to the caller -- which must mean the caller would own the reference, so it must be that we own the reference here. Which means if we return `NULL` and so _don't_ pass the pointer on to some other code that will own the reference, it's up to us here to decref it away.)
> 
> One is remove_importlib_frames -- great. A couple of other steps there don't appear to apply.

Am also returning null following a review from @serhiy-storchaka above. no?